File: 2023/trebuchet/part2.py
from re import findall

number_map = {
    "one": "1",
    "two": "2",
    "three": "3",
    "four": "4",
    "five": "5",
    "six": "6",
    "seven": "7",
    "eight": "8",
    "nine": "9",
}

lines: list[str] = []
with open("input.txt") as file:
    lines = file.readlines()

digits: list[int] = []
for line in lines:
    numbers_in_line = [
        number_map[num] if num in number_map else num
        for num in findall(r"\d|one|two|three|four|five|six|seven|eight|nine", line)
    ]

    if numbers_in_line:
        first_number = numbers_in_line[0]
        last_number = numbers_in_line[-1]

        digits.append(int(first_number + last_number))

print(sum(digits))

File: 2023/trebuchet/part2.js
const fs = require("fs");

const lines = fs.readFileSync("input.txt", "utf8").split(/\n/);

const numberMappings = {
  one: "1",
  two: "2",
  three: "3",
  four: "4",
  five: "5",
  six: "6",
  seven: "7",
  eight: "8",
  nine: "9",
};

let sum = 0;
for (const line of lines) {
  const regexLetterToNumber = Object.keys(numberMappings).join("|");
  const numbers = line.match(new RegExp(`\\d|${regexLetterToNumber}`, "g"));

  if (!numbers) {
    continue;
  }

  let firstNumber = numbers[0];
  let lastNumber = numbers[numbers.length - 1];

  if (firstNumber in numberMappings) {
    firstNumber = numberMappings[firstNumber];
  }

  if (lastNumber in numberMappings) {
    lastNumber = numberMappings[lastNumber];
  }

  sum += +firstNumber * 10 + +lastNumber;
}

console.log(sum);
